ibtk/src/utilities/CartGridPointwiseFunction.cpp

```diff
@@ -0,0 +1,275 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2023 - 2023 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+#include "ibtk/CartGridPointwiseFunction.h"
+
+#include <CartesianPatchGeometry.h>
+
+#include "ibtk/app_namespaces.h"
+
+namespace IBTK
+{
+/////////////////////////////// PUBLIC ///////////////////////////////////////
+template <typename F>
+CartGridPointwiseFunction<F>::CartGridPointwiseFunction(string object_name, F f)
+    : CartGridFunction(std::move(object_name)), d_f(f)
+{
+    // intentionally blank
+    return;
+} // CartGridPointwiseFunction
+
+template <>
+void
+CartGridPointwiseFunction<PointwiseFunctions::ScalarFcn>::setDataOnPatch(const int data_idx,
+                                                                         Pointer<Variable<NDIM> > var,
+                                                                         Pointer<Patch<NDIM> > patch,
+                                                                         const double data_time,
+                                                                         const bool initial_time,
+                                                                         Pointer<PatchLevel<NDIM> > /*patch_level*/)
+{
+    const Box<NDIM>& patch_box = patch->getBox();
+    const hier::Index<NDIM>& idx_low = patch_box.lower();
+    Pointer<CartesianPatchGeometry<NDIM> > pgeom = patch->getPatchGeometry();
+
+    const double* const xlow = pgeom->getXLower();
+    const double* const dx = pgeom->getDx();
+
+    // Set the data in the patch.
+    Pointer<PatchData<NDIM> > data = patch->getPatchData(data_idx);
+#if !defined(NDEBUG)
+    TBOX_ASSERT(data);
+#endif
+    Pointer<CellData<NDIM, double> > cc_data = data;
+    Pointer<NodeData<NDIM, double> > nc_data = data;
+    if (cc_data)
+    {
+        for (CellIterator<NDIM> ci(patch_box); ci; ci++)
+        {
+            const CellIndex<NDIM>& idx = ci();
+
+            VectorNd x;
+            for (int d = 0; d < NDIM; ++d) x[d] = xlow[d] + dx[d] * (static_cast<double>(idx(d) - idx_low(d)) + 0.5);
+            (*cc_data)(idx) = d_f((*cc_data)(idx), x, data_time);
+        }
```

Comment:
All of these functions have the same structure - we can get rid of a lot of duplication with some careful use of templates. Suppose we have an enumeration `DataCentering` for the type. We can do something like

```cpp
template <typename T> struct CenteringHelper {};

template <>
struct CenteringHelper<DataCentering::Side>
{
  static SideIterator<NDIM>
  get_iterator(const Box<NDIM> &box, const int component)
  {
    return SideIterator<NDIM>(box, component);
  }

  static VectorXd
  compute_physical_location(x_lower, box, refinement_ratio, component)
  {
    // something which computes the location given an index and patch lower bound, etc.
  }
};
```
where, internally, `component` is the same as depth for cell and node-centered but goes to `depth * NDIM` for the staggered layouts (we will need static helper functions for that too). For example, the cell-centered version of `compute_physical_location()` shouldn't use the component. With this we can write much more general loops of the sort

```cpp
template <DataCentering data_centering>
void assign(...)
{
  using Helper = CenteringHelper<data_centering>;
  auto it = Helper::get_iterator(box, component);
  for (; it; it++)
    {
      // works with all data centerings
      const auto &index = it();
      const VecNd x = Helper::compute_physical_location(index, ...);
      // etc.
    }
```


Even though it's more work I'd prefer to add these helpers so we can start cleaning these things up over time.

Reply:
I don't see how to make this work without writing new iterators for staggered fields. If we are acting on pointwise quantities as a whole (e.g. the entire `VectorNd` instead of components of the vector), then we don't loop through components, only through cells. We essentially want to compress the loops
```c++
for (CellIterator<NDIM> ci(box); ci; ci++)
{
    const CellIndex<NDIM>& idx = ci();
    // Do something on the cell
}

for (int axis = 0; axis < NDIM; ++axis)
{
    for (SideIterator<NDIM> si(box, axis); si; si++)
    {
         const SideIndex<NDIM>& idx = si();
         // Do something on the side
    }
}
```
into a single loop
```c++
using Helper = CenteringHelper<data_centering>
auto it = Helper::get_iterator(box)
for (; it; it++)
{
    // Do something on the index
}
```
For the loop over sides, we would need a new iterator to do this.